### DISCUSSION TITLE
Allow to define custom package manager commands

### DISCUSSION THREAD

**Discussion by ia3andy** [quarkiverse/quarkus-quinoa#133] [Ideas]
This discussion has been initiated by @computerlove in https://github.com/quarkiverse/quarkus-quinoa/pull/132.

I've been thinking into improving the customization of package manager commands in the long run (and not just "install").
I believe it would be nice to have something close to the script section in the package.json. e.g.:
```yarml
quarkus:
  quinoa:
  	package-manager-command:
       install-frozen: "npm ci --cache $CACHE_DIR/.npm --prefer-offline"
       // install: npm install (would still be the default)
       build-prod: "npm build-prod"
       build-dev: "npm build-dev"
       dev: "BROWSER=OFF npm dev" // default is "BROWSER=NONE npm start"

```

Instead of just the argument it would allow to edit the entire command (including envs) for one or more the available "actions". If not set if would keep the default (what we currently have).

To allow this, we might need to use something like [commons.apache.org/proper/commons-exec](https://commons.apache.org/proper/commons-exec/) and be careful about os compat.

Any thoughts?

**Comment by ia3andy** [comment]
@computerlove Would you like to take this on? It would be a very nice/big contribution!

**Reply by computerlove** [reply]
Yeah, I'll see what I can do :)

**Comment by gastaldi** [comment]
> Are the commands in `package.json` being replaced with this configuration? 

@gastaldi nop, package.json is just what's npm is using behind, but the goal with this is to allow any command (could possibly be something else than NodeJS).

> Why not change the `package.json` directly?

That was my initial idea, but it'd be nice to allow random project to work with Quinoa. Example if the UI project is owned by another team which doesn't want to change anything (could because of CI setup or whatever). 

This feature would allow any process that generate UI resources to be used alongside Quarkus.

**Comment by ia3andy** [comment]
It will also allow to disable commands `quarkus.quinoa.package-manager-command.install = echo 'NOOP'` (in case some tools might not need this particular command in their lifecycle)

**Comment by computerlove** [comment]
I think I will start on an approach with `quarkus.quinoa.package-manager-command.install`, `.build`, `.test`, `.dev` tomorrow. 
Having any setup for this in package.json seems a bit intrusive. I think this config belongs to Quinoa, so it feels more correct to place this in Quinoa's config.

**Reply by ia3andy** [reply]
That sounds good, I am currently evaluating (with the Quarkus team) using a library or not to exec the command. In the meantime, stick to the current stuff, maybe add `/bin/sh -c` on Linux and `cmd.exe /C` on Windows. Then we will adapt that part depending on what we decide. The goal would be to have something able to be called the same way as a script from the package.json. If you have any ideas on this, let us know :)

**Comment by computerlove** [comment]
I have a first draft ready: [feature-134](https://github.com/quarkiverse/quarkus-quinoa/compare/main...computerlove:quarkus-quinoa:feature/feature-134). 
It should be backwards compatible, allowing overriding the command and environment for each of `install`, `build`, `test`, `dev`.
```properties
quarkus.quinoa.package-manager-commands.install=npm install-all-packages --prefer-offline
quarkus.quinoa.package-manager-commands.install-env.MODE=secret
quarkus.quinoa.package-manager-commands.install-env.CI=maybe?
```
